presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayJoin.java

```diff
@@ -0,0 +1,206 @@
+                false, false);
+
+        @Override
+        public Signature getSignature()
+        {
+            return signature;
+        }
+
+        @Override
+        public boolean isHidden()
+        {
+            return false;
+        }
+
+        @Override
+        public boolean isDeterministic()
+        {
+            return true;
+        }
+
+        @Override
+        public String getDescription()
+        {
+            return DESCRIPTION;
+        }
+
+        @Override
+        public FunctionInfo specialize(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+        {
+            return specializeArrayJoin(types, functionRegistry, ImmutableList.of(false, false, false));
+        }
+
+    }
+
+    @Override
+    public Signature getSignature()
+    {
+        return SIGNATURE;
+    }
+
+    @Override
+    public boolean isHidden()
+    {
+        return false;
+    }
+
+    @Override
+    public boolean isDeterministic()
+    {
+        return true;
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return DESCRIPTION;
+    }
+
+    @Override
+    public FunctionInfo specialize(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        return specializeArrayJoin(types, functionRegistry, ImmutableList.of(false, false));
+    }
+
+    private static FunctionInfo specializeArrayJoin(Map<String, Type> types, FunctionRegistry functionRegistry, List<Boolean> nullableArguments)
+    {
+        Type type = types.get("T");
+        TypeSignature arrayType = parameterizedTypeName(StandardTypes.ARRAY, type.getTypeSignature());
+        Signature signature = new Signature(FUNCTION_NAME, VARCHAR_TYPE_SIGNATURE, arrayType, VARCHAR_TYPE_SIGNATURE, VARCHAR_TYPE_SIGNATURE);
+        MethodHandle methodHandle = createMethodHandle(type, functionRegistry);
+
+        return new FunctionInfo(signature, DESCRIPTION, false, methodHandle, true, false, nullableArguments);
+    }
+
+    private static MethodHandle createMethodHandle(Type type, FunctionRegistry functionRegistry)
+    {
+        FunctionInfo castFunction = functionRegistry.getExactFunction(internalOperator(CAST.name(), VARCHAR_TYPE_SIGNATURE, ImmutableList.of(type.getTypeSignature())));
+
+        if (!(type instanceof UnknownType)) {
+            checkCondition(castFunction != null, INVALID_FUNCTION_ARGUMENT, "Input type %s not supported", type);
+        }
+
+        MethodHandle methodHandle = methodHandle(ArrayJoin.class, "arrayJoin", FunctionInfo.class, Type.class, ConnectorSession.class, Slice.class, Slice.class, Slice.class);
+        return methodHandle.bindTo(castFunction).bindTo(type);
+    }
+
+    public static Slice arrayJoin(FunctionInfo castFunction, Type elementType, ConnectorSession session, Slice array, Slice delimiter, Slice nullReplacement) throws Throwable
```

Comment:
You need a second version of this (which can just delegate to this one) that doesn't have the third argument, for the version that has only two arguments